master/pkg/searcher/asha.go

```diff
@@ -1,169 +1,181 @@
-			expectedWorkloads += (cur.stepsNeeded - prev.stepsNeeded + 1) * cur.startTrials
-		} else {
-			expectedWorkloads += (cur.stepsNeeded + 1) * cur.startTrials
-		}
+				divisor:           config.Divisor,
+				stepsNeeded:       stepsNeeded,
+				startTrials:       startTrials,
+				outstandingTrials: 0})
 	}
 
 	return &asyncHalvingSearch{
 		AsyncHalvingConfig: config,
 		rungs:              rungs,
 		trialRungs:         make(map[RequestID]int),
 		earlyExitTrials:    make(map[RequestID]bool),
-		expectedWorkloads:  expectedWorkloads,
+		maxTrials:          config.MaxTrials,
 	}
 }
 
 type trialMetric struct {
 	requestID RequestID
 	metric    float64
+	promoted  bool
+	closed    bool
 }
 
 // rung describes a set of trials that are to be trained for the same number of steps.
 type rung struct {
-	stepsNeeded   int
-	metrics       []trialMetric
-	startTrials   int
-	promoteTrials int
+	divisor           float64
+	stepsNeeded       int
+	metrics           []trialMetric
+	startTrials       int
+	outstandingTrials int
+	promoteTrials     int
 }
 
 // promotions handles bookkeeping of validation metrics and returns a RequestID to promote if
 // appropriate.
-func (r *rung) promotions(requestID RequestID, metric float64) []RequestID {
+func (r *rung) promotionsAsync(requestID RequestID, metric float64) []RequestID {
+	// See if there is a trial to promote. We are increasing the total number of trials seen by 1; the
+	// number of best trials that definitely should have been promoted so far (numPromote) can only
+	// stay the same or increase by 1.
+	oldNumPromote := int(float64(len(r.metrics)) / r.divisor)
+	numPromote := int(float64(len(r.metrics)+1) / r.divisor)
+
 	// Insert the new trial result in the appropriate place in the sorted list.
 	insertIndex := sort.Search(
 		len(r.metrics),
 		func(i int) bool { return r.metrics[i].metric > metric },
 	)
+	promoteNow := insertIndex < numPromote
+
 	r.metrics = append(r.metrics, trialMetric{})
 	copy(r.metrics[insertIndex+1:], r.metrics[insertIndex:])
 	r.metrics[insertIndex] = trialMetric{
 		requestID: requestID,
 		metric:    metric,
+		promoted:  promoteNow,
+		closed:    false,
 	}
 
-	// If there are enough trials done to definitively promote one, do so. Otherwise, return nil.
-	currPromote := len(r.metrics) + r.promoteTrials - r.startTrials
-	switch {
-	case currPromote <= 0: // Not enough trials completed for any promotions.
-		return nil
-	case insertIndex < currPromote: // Incoming trial should be promoted.
+	// If the new trial is good enough, it should be promoted immediately (whether or not numPromote
+	// changes). Otherwise, if numPromote changes, there is some other trial that should be promoted,
+	// unless it has been promoted already.
+	if promoteNow {
 		return []RequestID{requestID}
-	default: // Promote next trial in sorted metrics array.
-		t := &r.metrics[currPromote-1]
-		return []RequestID{t.requestID}
 	}
+	if numPromote != oldNumPromote {
+		t := &r.metrics[oldNumPromote]
```

Comment:
question: I'm really not sure about this logic, but granted my understanding of this algorithm is limited. It seems like we should be promoting the right most element of the array that has not already been promoted.

Reply:
yes, we are doing that since things are zero indexed.  before trials from 0 to oldNumPromote-1 have been promoted so now the trial at index oldNumPromote needs to be promoted if possible.